### DISCUSSION TITLE
json-loader is still needed for webpack 5

### DISCUSSION THREAD

**Discussion by repolife** [webpack/webpack#14940] [General]
# Bug report

**What is the current behavior?**
According to documentation webpack 5 doesn't need json-loaders. 
https://webpack.js.org/migrate/3/#json-loader-is-not-required-anymore

**If the current behavior is a bug, please provide the steps to reproduce.**
import a local JSON file in a react component and try to use it.  You will receive an error 

`You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
`

**What is the expected behavior?**
JSON files should be read without need to install json-loader with webpack 5. 

**Other relevant information:**
webpack version:  ^5.57.1
Node.js version:  12.22.7
Operating System: Windows 10
Additional tools:

**Comment by alexander-akait** [comment]
You can use `asset/source` to load JSON as is

**Reply by alexander-akait** [reply]
I think you have invalid configuration if you get `You may need an appropriate loader to handle this file type`

**Reply by repolife** [reply]
I was able to resolve it by installing json-loader. 

`const HtmlWebPackPlugin = require("html-webpack-plugin");
const ModuleFederationPlugin = require("webpack/lib/container/ModuleFederationPlugin");

const deps = require("./package.json").dependencies;
module.exports = {
  output: {
    publicPath: "http://localhost:5050/",
  },

  resolve: {
    extensions: [".tsx", ".ts", ".jsx", ".js", ".json"],
  },

  devServer: {
    port: 5050,
    historyApiFallback: true,
  },

  module: {
    rules: [
      {
        test: /\.m?js/,
        type: "javascript/auto",
        resolve: {
          fullySpecified: false,
        },
      },
      {
        test: /\.(css|s[ac]ss)$/i,
        use: ["style-loader", "css-loader", "postcss-loader"],
      },
      {
        test: /\.(ts|tsx|js|jsx)$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
        },
      },
      {
        test: /\.json$/,
        loader: 'json-loader'
      }
    ],
  },

  plugins: [
    new ModuleFederationPlugin({
      name: "json_test",
      filename: "remoteEntry.js",
      remotes: {},
      exposes: {},
      shared: {
        ...deps,
        react: {
          singleton: true,
          requiredVersion: deps.react,
        },
        "react-dom": {
          singleton: true,
          requiredVersion: deps["react-dom"],
        },
      },
    }),
    new HtmlWebPackPlugin({
      template: "./src/index.html",
    }),
  ],
};`

> You can use `asset/raw` to load JSON as is

**Reply by repolife** [reply]
> You can use `asset/raw` to load JSON as is
I tried  the rule

`  {
        test: /\.json$/,
        type: 'asset/raw'
      }` 
      
      No Luck.

**Reply by alexander-akait** [reply]
Can you provide the whole error, with stack/file/etc?

**Reply by repolife** [reply]
`ERROR in ./package.json 2:8
Module parse failed: Unexpected token (2:8)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| {
>   "name": "json-test",
|   "version": "1.0.0",
|   "scripts": {
 @ ./src/App.tsx 3:0-39 9:320-328
 @ ./src/index.ts 1:0-15`
 
 App.tsx

`import React from "react";
import ReactDOM from "react-dom";
import * as env from "../package.json";

import "./index.css";

const App = () => (
  <div className="container">
    <div>Name: json-test</div>
    <div>Framework: react</div>
    <div>Language: TypeScript</div>
    <div>CSS: Empty CSS</div>
    <div>{env.name}</div>
  </div>
);
ReactDOM.render(<App />, document.getElementById("app"));`

**Reply by repolife** [reply]
> > You can use `asset/raw` to load JSON as is
> > I tried  the rule
> 
> ` { test: /\.json$/, type: 'asset/raw' }`
> 
> ```
>   No Luck. 
> ```

When using `{ test: /\.json$/, type: 'asset/raw' }`

I get: 
`Compiled with problems:X

ERROR in ./src/App.tsx 3:0-39

Module not found: Error: No parser registered for asset/raw`

**Reply by alexander-akait** [reply]
Please use `asset/source` https://webpack.js.org/guides/asset-modules/

**Reply by repolife** [reply]
> Please use `asset/source` https://webpack.js.org/guides/asset-modules/

I get this error now when using asset/raw. The JSON value doesn't show either. 

`Compiled with problems:X

WARNING in ./src/App.tsx 9:330-338

Can't import the named export 'name' (imported as 'env') from default-exporting module (only default export is available)
`

**Reply by alexander-akait** [reply]
It is mean you have invalid code, you should use `import something from "./file.json"`, JSON doesn't have named export

**Reply by repolife** [reply]
I'm importing the whole file.  I don't get this error when using json-loaders

```
import React from "react";
import ReactDOM from "react-dom";
import * as env from "../package.json";

import "./index.css";

const App = () => (
  <div className="container">
    <div>Name: json-test</div>
    <div>Framework: react</div>
    <div>Language: TypeScript</div>
    <div>CSS: Empty CSS</div>
    <div>name: {env.name}</div>
  </div>
);
ReactDOM.render(<App />, document.getElementById("app"));
```

**Reply by alexander-akait** [reply]
You can't use `import * as env from "../package.json";`, please use `import env from "../package.json";`

**Reply by repolife** [reply]
Yes, you can.  In typescript, it Imports all exports into the current module scope. I tried `import env from "../package.json`

The values don't show.

**Reply by sokra** [reply]
It's `type: "json"` for json, but that's the default for .json anyway...

**Reply by sokra** [reply]
The real problem is the rule with `test: /.m?js/,` which should be `test: /.m?js$/,` otherwise it matches .json too

**Reply by repolife** [reply]
I updated the rule as suggested. 

```
module: {
    rules: [
      {
        test: /\.m?js$/,
        type: "javascript/auto",
        resolve: {
          fullySpecified: false,
        },
      },
      {
        test: /\.(css|s[ac]ss)$/i,
        use: ["style-loader", "css-loader", "postcss-loader"],
      },
      {
        test: /\.(ts|tsx|js|jsx)$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
        },
      },
    
    ],
  },
```
  
  I'm getting error: `ERROR in ./src/App.tsx 3:0-39
Module not found: Error: No parser registered for asset/json
 @ ./src/index.ts 1:0-15
`

**Reply by alexander-akait** [reply]
There is not `asset/json` type, remove `asset`, only `type: "json"` exists

**Reply by repolife** [reply]
```
      { test: /\.json$/, type: 'json' }
```

Worked. I hope the documentation gets updated to provide these instructions as it makes seem that nothing really needs to be done for JSON files.

**Reply by sokra** [reply]
There is nothing needed for .json. The rule you pasted is the default.

**Reply by sushruth** [reply]
I can confirm that without using what @alexander-akait and @repolife suggested is the only solution that worked for me with webpack 5.68 for my setup.
```
{ test: /\.json$/, type: 'json' }
```
Works!

**Comment by ozkey** [comment]
this works for me:    

 `      { test: /\.json$/, type: 'json' },`

**Comment by fabb** [comment]
I needed to update `resolve.extensions` to include `.json`:

```
        resolve: {
            extensions: ['.ts', '.tsx', '.js', '.jsx', '.json'],
        },
```
